sys/linux/dev_qat_adf_ctl.txt

```diff
@@ -0,0 +1,108 @@
+
+include <linux/fcntl.h>
+include <drivers/crypto/qat/qat_common/adf_cfg_user.h>
+include <drivers/crypto/qat/qat_common/adf_cfg_common.h>
+resource fd_qat[fd]
+
+openat$qat_adf_ctl(fd const[AT_FDCWD], file ptr[in, string["/dev/qat_adf_ctl"]], flags flags[open_flags], mode const[0]) fd_qat
+ioctl$IOCTL_CONFIG_SYS_RESOURCE_PARAMETERS(fd fd_qat, cmd const[IOCTL_CONFIG_SYS_RESOURCE_PARAMETERS], arg ptr[in, adf_user_cfg_ctl_data])
+ioctl$IOCTL_STOP_ACCEL_DEV(fd fd_qat, cmd const[IOCTL_STOP_ACCEL_DEV], arg ptr[in, adf_user_cfg_ctl_data])
+ioctl$IOCTL_START_ACCEL_DEV(fd fd_qat, cmd const[IOCTL_START_ACCEL_DEV], arg ptr[in, adf_user_cfg_ctl_data])
+ioctl$IOCTL_GET_NUM_DEVICES(fd fd_qat, cmd const[IOCTL_GET_NUM_DEVICES], arg ptr[out, int32])
+ioctl$IOCTL_STATUS_ACCEL_DEV(fd fd_qat, cmd const[IOCTL_STATUS_ACCEL_DEV], arg ptr[inout, adf_dev_status_info])
+
+adf_user_cfg_ctl_data {
+	config_section	adf_ptr64[adf_user_cfg_section]
+	device_id	int8
+} [packed]
+
+# Section
+
+adf_user_cfg_section {
+	name	string[qat_section_name, ADF_CFG_MAX_SECTION_LEN_IN_BYTES]
+	params	adf_ptr64[adf_user_cfg_key_val]
+	next	adf_ptr64[adf_user_cfg_section]
+} [packed]
+
+qat_section_name = "GENERAL", "KERNEL", "Accelerator", "Accelerator0", "Accelerator1"
+
+# Key-values
+
+adf_user_cfg_key_val [
+	adf_dec	adf_user_cfg_key_val_t[ADF_DEC, adf_user_cfg_val_dec]
+	adf_hex	adf_user_cfg_key_val_t[ADF_HEX, adf_user_cfg_val_hex]
+	adf_str	adf_user_cfg_key_val_t[ADF_STR, adf_user_cfg_val_str]
+]
+
+type adf_user_cfg_key_val_t[TYPE, VAL] {
+	key	adf_user_cfg_key
+	val	VAL
+	next	adf_ptr64[adf_user_cfg_key_val]
+	type	const[TYPE, int32]
+} [packed]
+
+# Key
+
+adf_user_cfg_key [
+	normal	string[adf_user_cfg_normal_keys, ADF_CFG_MAX_KEY_LEN_IN_BYTES]
+	format	adf_user_cfg_format_keys
+]
+
+adf_user_cfg_normal_keys = "NumberCyInstances", "NumberDcInstances"
+
+adf_user_cfg_format_keys {
+	prefix	stringnoz[adf_user_cfg_format_keys_prefix]
+	number	stringnoz[adf_user_cfg_format_keys_numbers]
+	postfix	string[adf_user_cfg_format_keys_postfix, ADF_KEYS_POSTFIX_MAX_LEN]
+} [size[ADF_CFG_MAX_KEY_LEN_IN_BYTES]]
+
+adf_user_cfg_format_keys_prefix = "Cy", "Dc"
+adf_user_cfg_format_keys_numbers = "0", "1", "2", "3", "4", "5"
+adf_user_cfg_format_keys_postfix = "NumConcurrentSymRequests", "NumConcurrentAsymRequests", "NumConcurrentRequests", "RingAsymTx", "RingSymTx", "RingAsymRx", "RingSymRx", "RingTx", "RingRx", "BankNumber"
+
+# This constant 30 needs to be at least the length of the longest string in adf_user_cfg_format_keys_postfix
+define ADF_KEYS_POSTFIX_MAX_LEN	30
+
+# Value
+
+adf_user_cfg_val_dec {
+	val	fmt[dec, intptr]
+	null	const[0, int8]
+} [packed, size[ADF_CFG_MAX_VAL_LEN_IN_BYTES]]
+
+adf_user_cfg_val_hex {
+	val	fmt[hex, intptr]
+	null	const[0, int8]
+} [packed, size[ADF_CFG_MAX_VAL_LEN_IN_BYTES]]
+
+adf_user_cfg_val_str {
+	val	array[int8, ADF_CFG_MAX_VAL_LEN_IN_BYTES]
+}
+
+# Ptr
+
+type adf_ptr64[T] {
```

Comment:
General Comment: I think this type works properly but it is easier to understand the code if you follow source code pattern. For example, instead of defining `type adf_ptr64[ T= adf_user_ctl_data] ` , you may prefer to define a union as following 
```
adf_user_ctl_data[ 
        cfg_section_padding cfg_section_padding
        padding int64
] 
``` 
where `cfg_section_padding` defined as a union as well.

Reply:
The padding field is never used and only exists to force the pointer to be 8 bytes. Also, I don't think we can use `ptr64` here because it will be wrong in different endianess machine.
There are 4 places that uses `adf_ptr64`, so if we expand those definition, it will create 4 new struct, which I don't think that will make the code easier to follow. I will ping other reviewer @jedav @morehouse @xairy and see what they think. Thanks!

Reply:
If its only use to force the pointer to be 8 bytes, I think you should use ptr64.

Reply:
I don't think that will be correct in big-endian 32 bit machines. In this case an 32 bit pointer that is stored in a `ptr64` would have layout
```
padding int32
ptr ptr32
```
which is different then the layout `adf_user` uses. (Or because this is an intel thing it can only be used on little-endian machines?)

Reply:
Yeah, what they do here is weird. I'd say @bobogei81123 is right, the pointer should always be stored in the first 4 bytes for 32-bit arches regardless of the endianness, which isn't the case for `ptr64`. Let's keep the `adf_ptr64 ` template.

Reply:
I will resolve this issue for now. Feel free to unresolve if there are further questions.